File: cpp/1046.0_Last_Stone_Weight.cpp
/*
max heap + simulation
T: O(N + NlogN)
S: O(N)

Runtime: 4 ms, faster than 37.54% of C++ online submissions for Last Stone Weight.
Memory Usage: 7.6 MB, less than 78.96% of C++ online submissions for Last Stone Weight.
*/
class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {
        priority_queue<int> pq(stones.begin(), stones.end());
        while (pq.size() > 1) {
            int first = pq.top();
            pq.pop();
            int second = pq.top();
            pq.pop();
            if (first != second) {
                pq.push(first - second);
            }
        }
        return pq.empty() ? 0: pq.top();
    }
};

File: java/1046.0_Last_Stone_Weight.java
/*
T: O((N^2)*logN)

执行用时：0 ms, 在所有 Java 提交中击败了100.00% 的用户
内存消耗：38.9 MB, 在所有 Java 提交中击败了32.61% 的用户
通过测试用例：70 / 70
*/
class Solution {
    public int lastStoneWeight(int[] stones) {
        int n = stones.length;
        if (n == 2)
            return Math.abs(stones[0] - stones[1]);
        if (n == 1)
            return stones[0];
        Arrays.sort(stones);
        if (stones[n - 3] == 0)
            return stones[n - 1] - stones[n - 2];
        stones[n - 1] -= stones[n - 2];
        stones[n - 2] = 0;
        return lastStoneWeight(stones);
    }
}
